File: Programmers/Lv.1/가장 많이 받은 선물/258712.js
function solution(friends, gifts) {
    let answer = 0;
    
    const indexObj = {};
    const countArr = [], giftArr = [];
    friends.forEach((v, i) => {
        indexObj[v] = i;
        countArr.push(0);
        giftArr.push(Array(friends.length).fill(0));
    });
    
    gifts.forEach(v => {
        const [g1, g2] = v.split(' ');
        countArr[indexObj[g1]]++;
        countArr[indexObj[g2]]--;
        giftArr[indexObj[g1]][indexObj[g2]]++;
    });
    
    for (let i = 0; i < giftArr.length; i++) {
        let count = 0;
        
        for (let j = 0; j < giftArr.length; j++) {
            if (i === j) continue;
            
            if (giftArr[i][j] > giftArr[j][i]) {
                count++;
            } else if (giftArr[i][j] === giftArr[j][i]) {
                if (countArr[i] > countArr[j]) {
                    count++;
                }
            }
            
            answer = Math.max(answer, count);
        }
    }
    
    return answer;
}

File: Programmers/Lv.1/가장 많이 받은 선물/258712.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        
        Map<String, Map<String, Integer>> giftMap = new HashMap<>();
        Map<String, Integer> countMap = new HashMap<>();
        for (String friend : friends) {
            Map<String, Integer> map = new HashMap<>();
            giftMap.put(friend, map);
            countMap.put(friend, 0);
        }
        
        for (String gift : gifts) {
            String[] g = gift.split(" ");
            Map<String, Integer> gMap = giftMap.get(g[0]);
            gMap.put(g[1], gMap.getOrDefault(g[1], 0) + 1);
            countMap.put(g[0], countMap.get(g[0]) + 1);
            countMap.put(g[1], countMap.get(g[1]) - 1);
        }
        
        int[] nextMonth = new int[friends.length];
        for (int i = 0; i < friends.length - 1; i++) {
            Map<String, Integer> map = giftMap.get(friends[i]);
            for (int j = i + 1; j < friends.length; j++) {
                int num1 = map.getOrDefault(friends[j], 0);
                int num2 = giftMap.get(friends[j]).getOrDefault(friends[i], 0);
                if (num1 > num2) {
                    nextMonth[i]++;
                } else if (num1 < num2) {
                    nextMonth[j]++;
                } else if (num1 == num2) {
                    int cnt1 = countMap.get(friends[i]);
                    int cnt2 = countMap.get(friends[j]);
                    if (cnt1 > cnt2) {
                        nextMonth[i]++;
                    } else if (cnt2 > cnt1) {
                        nextMonth[j]++;
                    }
                }
            }
        }
        
        Arrays.sort(nextMonth);
        
        answer = nextMonth[nextMonth.length - 1];
        
        return answer;
    }
}
